internal/pubsub/drivertest/drivertest.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2018 The Go Cloud Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package drivertest provides a conformance test for implementations of
+// driver.
+package drivertest
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"testing"
+
+	"github.com/google/go-cloud/internal/pubsub"
+)
+
+// Harness descibes the functionality test harnesses must provide to run
+// conformance tests.
+type Harness interface {
+	// MakeTopicDriver creates a Topic and associated Subscription to test.
+	MakePair() (*pubsub.Topic, *pubsub.Subscription, error)
+
+	// Close closes resources used by the harness.
+	Close()
+}
+
+// HarnessMaker describes functions that construct a harness for running tests.
+// It is called exactly once per test; Harness.Close() will be called when the test is complete.
+type HarnessMaker func(ctx context.Context, t *testing.T) (Harness, error)
+
+// RunConformanceTests runs conformance tests for provider implementations of pubsub.
+func RunConformanceTests(t *testing.T, newHarness HarnessMaker) {
+	t.Run("TestSendReceive", func(t *testing.T) {
+		testSendReceive(t, newHarness)
+	})
+	t.Run("TestSendError", func(t *testing.T) {
+		testSendError(t, newHarness)
+	})
+	t.Run("TestReceiveError", func(t *testing.T) {
+		testReceiveError(t, newHarness)
+	})
+	t.Run("TestCancelSendReceive", func(t *testing.T) {
+		testCancelSendReceive(t, newHarness)
+	})
+	t.Run("TestCancelAck", func(t *testing.T) {
+		testCancelAck(t, newHarness)
+	})
+}
+
+// testSendReceive tests that a single message sent to a Topic gets received
+// from a corresponding Subscription.
+func testSendReceive(t *testing.T, newHarness HarnessMaker) {
+	ctx := context.Background()
+	h, err := newHarness(ctx, t)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer h.Close()
+	top, sub, err := h.MakePair()
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Send to the topic.
+	m := &pubsub.Message{
+		Body:     []byte(randStr()),
+		Metadata: map[string]string{randStr(): randStr()},
+	}
+	if err := top.Send(ctx, m); err != nil {
+		t.Fatal(err)
+	}
+
+	// Receive from the subscription.
+	m2, err := sub.Receive(ctx)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// Check that the received message matches the sent one.
```

Comment:
Use `cmp.Diff` on the message objects? Or at least on `.Metadata`.

Reply:
Done.